### PR TITLE
Vmo 4152 resource validation

### PR DESCRIPTION
**Pull request by safydy-r**
This PR implements main `resource validation`, except for:
- validation on expressions (https://viamoinc.atlassian.net/browse/VMO-5190)
- deep linking to select the right language tabs (https://viamoinc.atlassian.net/browse/VMO-5191)
- Update resources when we changes the activeFlow's languages (https://viamoinc.atlassian.net/browse/VMO-5213)

Merge plan:
- merge flow-runner PR first: https://github.com/FLOIP/flow-runner/pull/59
- update flow-builder package.json, replace `"@floip/flow-runner": "https://github.com/FLOIP/flow-runner.git#VMO-4152-resource-validation",` to `"@floip/flow-runner": "https://github.com/FLOIP/flow-runner.git#feature/VMO-4281-flow-spec-rc2`
- merge this PR

[Overview]
vmo-4152

### CODE REVIEW

File: package.json
Comment (safydy-r) [inline package.json:26]: Will revert after merging https://github.com/FLOIP/flow-runner/pull/59

File: src/components/interaction-designer/resource-editors/ResourceVariantTextEditor.vue
Comment (safydy-r) [inline src/components/interaction-designer/resource-editors/ResourceVariantTextEditor.vue:2]: Just wrapping everything within `<validation-message`

File: src/components/interaction-designer/toolbar/BlockErrorsExpandable.vue
Comment (safydy-r) [inline src/components/interaction-designer/toolbar/BlockErrorsExpandable.vue:51]: @bulatgab , FYI here, I increased this as we have more validation errors to display

File: src/components/interaction-designer/toolbar/ErrorNotifications.vue
Comment (safydy-r) [inline src/components/interaction-designer/toolbar/ErrorNotifications.vue:110]: When thinking about this block of code, I noticed we can remove the mapping because block's validation are already prefixed with `block/`

File: src/components/interaction-designer/toolbar/BlockErrorsExpandable.vue
Comment (bulatgab) [reply src/components/interaction-designer/toolbar/BlockErrorsExpandable.vue:51]: Sounds good!

**Comment by bulatgab** [comment]
The jsdoc for `flattenErrorMessages()` (in `flow-builder/src/store/validation/index.ts`) has examples with flow and block errors. Could you add an example of a resource error to that list as well, please?

File: src/components/interaction-designer/resource-editors/ResourceEditor.vue
Comment (bulatgab) [inline src/components/interaction-designer/resource-editors/ResourceEditor.vue:242]: As we discussed, `computeResourceIndex` is currently fragile (if we change the number of modes for the flow).
It would be great to find a way to make it more robust.

File: src/components/interaction-designer/toolbar/BlockErrorsExpandable.vue
Comment (bulatgab) [reply src/components/interaction-designer/toolbar/BlockErrorsExpandable.vue:46]: minor eslint issues like this can be fixed

File: src/components/interaction-designer/toolbar/BlockErrorsExpandable.vue
Comment (bulatgab) [reply src/components/interaction-designer/toolbar/BlockErrorsExpandable.vue:72]: minor eslint issues like this can be fixed

File: src/components/interaction-designer/toolbar/BlockErrorsExpandable.vue
Comment (bulatgab) [inline src/components/interaction-designer/toolbar/BlockErrorsExpandable.vue:72]: The block validation errors are passed down from the parent component as a 'status' _prop_, whereas resource validation errors are read from the _vuex store_.
Do you think it would make sense to unify this and read the block validation errors from the store as well?

P.S. I think I was the one who did it via the props, ignoring the global store approach. My bad 😄

File: src/components/interaction-designer/resource-editors/ResourceEditor.vue
Comment (safydy-r) [reply src/components/interaction-designer/resource-editors/ResourceEditor.vue:242]: @bulatgab , thanks for raising flag here. I do agree using number `index` here could be confusing, but that's how the jsonSchema validate values of array (`resources.values` is an array).
I spent time to find alternative options, but couldn't so far.
Fortunately, I was able to correct the resources.values when users update `modes`; but it gets a bit tricky when they update languages. 

Therefore, I created a new ticket to handle that: https://viamoinc.atlassian.net/browse/VMO-5213

File: src/components/interaction-designer/toolbar/BlockErrorsExpandable.vue
Comment (safydy-r) [reply src/components/interaction-designer/toolbar/BlockErrorsExpandable.vue:72]: @bulatgab, I think we don't really need to always get data from vuex (or always send them from parent).
The key point here: As`BlockErrorsExpandable.vue` is handling errors for **01 block**. So, if I change `resourceValidationStatuses` into `resourceValidationStatusesForCurrentBlockValidationStatus`, that would be less confusing. What do you think ?

File: src/components/interaction-designer/flow-editors/FlowEditor.vue
Comment (safydy-r) [inline src/components/interaction-designer/flow-editors/FlowEditor.vue:83]: @bulatgab : I made update here after your 1st round review

File: src/components/interaction-designer/flow-editors/FlowEditor.vue
Comment (safydy-r) [inline src/components/interaction-designer/flow-editors/FlowEditor.vue:91]: @bulatgab : I made update here after your 1st round review

File: src/components/interaction-designer/flow-editors/LanguagesEditor.vue
Comment (safydy-r) [inline src/components/interaction-designer/flow-editors/LanguagesEditor.vue:39]: @bulatgab : I made update here after your 1st round review

File: src/store/flow/flow.ts
Comment (safydy-r) [inline src/store/flow/flow.ts:126]: @bulatgab : I made update here after your 1st round review

File: src/store/flow/flow.ts
Comment (safydy-r) [inline src/store/flow/flow.ts:133]: @bulatgab : I made update here after your 1st round review

File: src/components/interaction-designer/toolbar/BlockErrorsExpandable.vue
Comment (bulatgab) [reply src/components/interaction-designer/toolbar/BlockErrorsExpandable.vue]: Okay, a mixed approach then.
Regarding the `resourceValidationStatuses` - personally, it wasn't confusing to me, but I'm okay with the new name.

File: src/components/interaction-designer/resource-editors/ResourceEditor.vue
Comment (jcbashdown) [reply src/components/interaction-designer/resource-editors/ResourceEditor.vue:242]: I also have some concerns here. It looks like you are making the index of resource language modes predictable by sorting when languages and modes change and then using the predictability to look up using an integer index. This is going to be fragile whatever I think - what if someone forgets to sort in future in a new place these values are changed? It also doesn't seem to work as intended though as the resource languages order doesn't seem to match this sorted order?

I understand this may be fixed in 5213. Something we might want to consider then is running a separate validation for values outside of the full resource validation. That way we can validate more specifically based on the content type (e.g. only allow a file path for certain content types) and we can also key the validation errors however we want rather than relying on the structure of the resource. For example, we can have string keys based on the resource uuid/mode/language id combination rather than relying on the order of an array. Does this make sense? Maybe this would be worse as it would involve a pretty different approach to what we use elsewhere but maybe also worth considering to remove the sort and resolve other issues here?

**Review by jcbashdown — CHANGES_REQUESTED** [review]
Looks good. Just a couple of things that might need changing and I think there are some linting errors on lines you've changed to address as well. Then in the tickets you've spun out we might also want to do a refactor to make things more robust to changes languages (and modes).

File: src/components/interaction-designer/resource-editors/ResourceEditor.vue
Comment (safydy-r) [reply src/components/interaction-designer/resource-editors/ResourceEditor.vue:242]: @jcbashdown , thanks for input here.
My main blocker is that if we validate an array using AJV, the returned corresponding `dataPath` is `numeric indexed`, see this vuex state for eg:
Screen Shot 2022-01-25 at 3 49 22 PM

The numeric index corresponds to the order of the Data in the array, that's why it's important to follow a certain order. I know this is quite fragile, and I'm open if we can find alternative. The only solution I found so far:
- force order (like I did in this PR)
- change the resources spec from `array` to object

Any suggestion?

File: src/components/interaction-designer/toolbar/BlockErrorsExpandable.vue
Comment (safydy-r) [reply src/components/interaction-designer/toolbar/BlockErrorsExpandable.vue:86]: Thanks, this would be better I think. I changed it !

File: src/components/interaction-designer/toolbar/ErrorNotifications.vue
Comment (safydy-r) [reply src/components/interaction-designer/toolbar/ErrorNotifications.vue:110]: @jcbashdown , nice catch here, I think you're right !
I reverted by changes and added a doc string to the getter to make sure we understand that mapping part.

File: src/components/interaction-designer/resource-editors/ResourceEditor.vue
Comment (jcbashdown) [reply src/components/interaction-designer/resource-editors/ResourceEditor.vue:242]: The issue I was finding - the actual ordering of the language values in the resource on your branch are not consistent. They still depend on the order in which I add the language. That means I can add the second language in the list on creation and then go back and add the first. Now when I edit the second tab in the UI the first becomes valid and vice versa.

This issue isn't quite the same as the one in VMO-5213 as that's about the fact that existing resources don't get new empty values when you add a new language.

The solution I'm proposing for the issue I've found here is that we are probably going to need to validate resource value strings directly and build up our own nested structure of validation errors corresponding to the IDs of language and modes. This is much more robust to edits of languages and modes than what we have but sacrifices the convenience of using the json schema for resources. It has the advantage though, of potentially letting us create specific validators for each content type.

Let's try and have a call at some point to discuss further.

**Review by jcbashdown — APPROVED** [review]

**Comment by jcbashdown** [comment]
I'm approving this pending 3 things:
1) Resolution of the conflict + merging in master before QA
2) Fixing of linting errors on new/edited lines
3) A quick spike on fixing the problems with validating (and displaying validation errors) when languages are added and removed. We agreed in a call to try and fix this by trying to add values to resources immediately when this happens and by sorting within resources.

If 3 proves complicated then we will do this in it's own ticket - VMO-5213

File: src/components/interaction-designer/resource-editors/ResourceEditor.vue
Comment (safydy-r) [reply src/components/interaction-designer/resource-editors/ResourceEditor.vue:242]: As discussed in our call, we will handle languages related behaviour in https://viamoinc.atlassian.net/browse/VMO-5213

**Pull request merged**
